Compiler/src/abstractinterpretation.jl

```diff
@@ -3396,28 +3396,31 @@ function abstract_eval_globalref(interp::AbstractInterpreter, g::GlobalRef, sv::
     return ret
 end
 
-function global_assignment_exct(interp::AbstractInterpreter, sv::AbsIntState, g::GlobalRef, @nospecialize(newty))
+function global_assignment_rt_exct(interp::AbstractInterpreter, sv::AbsIntState, g::GlobalRef, @nospecialize(newty))
     partition = abstract_eval_binding_partition!(interp, g, sv)
-    return global_assignment_binding_exct(partition, newty)
+    return global_assignment_binding_rt_exct(partition, newty)
 end
 
-function global_assignment_binding_exct(partition::Core.BindingPartition, @nospecialize(newty))
+function global_assignment_binding_rt_exct(partition::Core.BindingPartition, @nospecialize(newty))
     kind = binding_kind(partition)
-    if is_some_guard(kind) || is_some_const_binding(kind)
-        return ErrorException
+    if is_some_guard(kind)
+        return Pair{Any,Any}(newty, ErrorException)
+    elseif is_some_const_binding(kind)
+        return Pair{Any,Any}(Union{}, ErrorException)
     end
-
     ty = partition_restriction(partition)
-    if !(widenconst(newty) <: ty)
-        return TypeError
+    wnewty = widenconst(newty)
+    if !hasintersect(wnewty, ty)
+        return Pair{Any,Any}(Bottom, TypeError)
+    elseif !(wnewty <: ty)
+        return Pair{Any,Any}(newty, TypeError)
```

Comment:
use the typeintersect_tfunc here to compute newty most precisely?

Reply:
Do you mean `tmeet(𝕃, newty, ty)`?

Reply:
Yes, I guess that is true. Often the types are in the lifted domain, but in this case the partition_restriction is already in the value domain, so the instanceof_tfunc isn't necessary